express-delete/server.ts

```diff
@@ -0,0 +1,54 @@
+import express from 'express';
+
+type Grade = {
+  id: number;
+  name: string;
+  course: string;
+  score: number;
+};
+
+const grades: Record<number, Grade> = {
+  12: {
+    id: 12,
+    name: 'Tim Berners-Lee',
+    course: 'HTML',
+    score: 95,
+  },
+  47: {
+    id: 47,
+    name: 'Brendan Eich',
+    course: 'JavaScript',
+    score: 100,
+  },
+  273: {
+    id: 273,
+    name: 'Håkon Wium Lie',
+    course: 'CSS',
+    score: 92,
+  },
+};
+
+const app = express();
+
+app.get('/api/grades', (req, res) => {
+  const gradesArray: Grade[] = [];
+  for (const prop in grades) {
+    gradesArray.push(grades[prop]);
+  }
+  res.json(gradesArray);
+});
+
+app.listen(8080, () => {
+  console.log('Express server listening on port 8080');
+});
+
+app.delete('/api/grades/:id', (req, res) => {
+  const gradeId = req.params.id;
+
+  delete grades[gradeId];
+  res.sendStatus(204);
+  if (!gradeId) {
+    res.sendStatus(404);
```

Comment:
Oops, you can't do it this way. Here's what happens if they specify an id that doesn't exist:
1. you get the id and assign it to `gradeId`
2. you try to delete `grades[gradeId]`
3. You send a 204
4. You check to see if `gradeId` is defined.  (It will probably be defined and be equal to the `id` param.)
5. you send a 404, but you've already sent a 204 so you'll get an Express error that you are sending 2 responses.

Instead, you need to check if the grade exists (not if the gradeId is defined). If it does, delete it and send a 204. If it does not, send a 404. And make sure that you don't accidentally send 2 responses.